stl/inc/syncstream

```diff
@@ -0,0 +1,360 @@
+// syncstream standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _SYNCSTREAM_
+#define _SYNCSTREAM_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+#if !_HAS_CXX20
+#pragma message("The contents of <syncstream> are available only with C++20 or later.")
+#else // ^^^ !_HAS_CXX20 / _HAS_CXX20 vvv
+#include <memory>
+#include <ostream>
+#include <shared_mutex>
+#include <streambuf>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#pragma push_macro("emit")
+#undef new
+#undef emit
+
+_STD_BEGIN
+
+extern "C" shared_mutex* __cdecl _Get_mutex_for_instance(void* _Ptr) noexcept;
+extern "C" void __cdecl _Acquire_mutex_for_instance(void* _Ptr);
+extern "C" void __cdecl _Release_mutex_for_instance(void* _Ptr) noexcept;
+
+// CLASS TEMPLATE _Basic_osyncbuf_impl
+template <class _Elem, class _Traits>
+class _Basic_syncbuf_impl : public basic_streambuf<_Elem, _Traits> {
+public:
+    void _Set_emit_on_sync(const bool _Val) noexcept {
+        _Emit_on_sync = _Val;
+    }
+
+    virtual bool _Do_emit() = 0;
+
+protected:
+    using _Mysb = basic_streambuf<_Elem, _Traits>;
+
+    _Basic_syncbuf_impl() = default;
+
+    _Basic_syncbuf_impl(_Basic_syncbuf_impl&& _Right) {
+        swap(_Right);
+    }
+
+    void swap(_Basic_syncbuf_impl& _Right) {
+        if (this != _STD addressof(_Right)) {
+            _Mysb::swap(_Right);
+            _STD swap(_Right._Emit_on_sync, _Emit_on_sync);
+        }
+    }
+
+    bool _Emit_on_sync{false};
+};
+
+// CLASS TEMPLATE basic_syncbuf
+template <class _Elem, class _Traits, class _Alloc>
+class basic_syncbuf : public _Basic_syncbuf_impl<_Elem, _Traits> {
+public:
+    using int_type       = typename _Traits::int_type;
+    using pos_type       = typename _Traits::pos_type;
+    using off_type       = typename _Traits::off_type;
+    using allocator_type = _Alloc;
+    using streambuf_type = basic_streambuf<_Elem, _Traits>;
+
+    using _Mybase   = _Basic_syncbuf_impl<_Elem, _Traits>;
+    using pointer   = typename allocator_traits<_Alloc>::pointer;
+    using size_type = typename allocator_traits<_Alloc>::size_type;
+
+    basic_syncbuf() = default;
+
+    explicit basic_syncbuf(streambuf_type* _Strbuf) : basic_syncbuf(_Strbuf, _Alloc{}) {}
+
+    basic_syncbuf(streambuf_type* _Strbuf, const _Alloc& _Al_) : _Wrapped(_Strbuf), _Al(_Al_) {
+        if (_Wrapped) {
+            _Acquire_mutex_for_instance(_Wrapped);
+        }
+        _Init();
+    }
+
+    basic_syncbuf(basic_syncbuf&& _Right) {
+        _Al = _STD move(_Right._Al);
+        _Assign_rv(_STD move(_Right));
+    }
+
+    ~basic_syncbuf() {
+        _Emit();
+        _Tidy();
+    }
+
+    basic_syncbuf& operator=(basic_syncbuf&& _Right) {

```

Comment:
This function is required by the standard to be `noexcept`.

Reply:
This is confusing to me, the class template does not have `noexcept` but later it's marked as `noexcept`.
Also, `_Move_assign` with `_No_propagate_allocators` could call `allocate` which might throw.

Reply:
Yep, this looks like it should be reported as an LWG issue - `noexcept` appears to be incorrect in the Standard.

Reply:
I found LWG-3498 reporting this issue, but it has not been voted yet

Reply:
Added comments citing LWG-3498. Thanks!